src/librustc/hir/lowering.rs

```diff
@@ -5546,16 +5546,26 @@ impl<'a> LoweringContext<'a> {
         params: Option<P<hir::GenericArgs>>,
         is_value: bool,
     ) -> hir::Path {
-        let mut path = self.resolver
+        let (path, res) = self.resolver
             .resolve_str_path(span, self.crate_root, components, is_value);
-        path.segments.last_mut().unwrap().args = params;
 
-        for seg in path.segments.iter_mut() {
-            if seg.hir_id.is_some() {
-                seg.hir_id = Some(self.next_id());
+        let mut segments: Vec<_> = path.segments.iter().map(|seg| {
+            let mut hir_seg = hir::PathSegment::from_ident(seg.ident);
+            if hir_seg.hir_id.is_some() {
```

Comment:
This is never true. `from_ident` just fills the `hir_id` field with `None`

Reply:
From all that I can tell this was also dead code before this PR

Reply:
Yes but... I'd rather keep the dead code? IMO path segments should either always have a `HirId`, or never, not "sometimes".

Reply:
>IMO path segments should either always have a HirId, or never, not "sometimes".

I... don't understand the connection with the dead code.

Reply:
change it to an assert then?

Reply:
@petrochenkov @oli-obk When IDs were added to path segments, they were made optional in (at least) the HIR. I'd prefer if they weren't optional, then this wouldn't be dead code but actually assign IDs - and, well, it should be correlating those `HirId`s with the `NodeId`s in the AST path segments).

The dead code has been dead all along, and presumably is sort of "generic" over whether `hir::PathSegment::from_ident` uses `None` or `Some(DUMMY_HIR_ID)` (although I'm not sure that's even a thing).

I'd rather make the currently-dead code unconditional, and relate the created `HirId`s with `seg.id`, if you wouldn't mind that?

Reply:
I agree that we should eagerly generate ids here, but since this PR is a blocker for other PRs I think we should start out with not changing behaviour. So maybe just remove the entire `if` statement and put a FIXME there and do the `HirId` generation change in a separate PR?

Reply:
I guess I'll look at how `ast::PathSegment` is lowered normally in `hir::lowering` and then copy that here.